<?xml version='1.0' encoding='utf-8'?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="my_robot_arm">

    <xacro:include filename="$(find magni_description)/urdf/inertial.xacro" />
    <xacro:property name="initial_positions" value="${load_yaml(find(my_robot_arm)/config)['initial_positions']}"/>

    <!-- Include a UR5 -->
    <xacro:include filename="$(find ur_description)/urdf/ur_macro.xacro" />
    <xacro:arg name="ur_type" default="ur5e"/>

    <xacro:ur_robot name="ur5" tf_prefix="ur5_" parent="world" joint_limits_parameters_file="$(find ur_description)/config/$(arg ur_type)/joint_limits.yaml" kinematics_parameters_file="$(find ur_description)/config/$(arg ur_type)/default_kinematics.yaml" physical_parameters_file="$(find ur_description)/config/$(arg ur_type)/physical_parameters.yaml" visual_parameters_file="$(find ur_description)/config/$(arg ur_type)/visual_parameters.yaml">
        <origin xyz="-0.1 0 0.15" rpy="${pi} 0 0" />
    </xacro:ur_robot>
    <link name="world" />

    <link name="base_link">
        <visual>
            <!--Base_Link_Part/Body.-->
            <origin rpy="0.0 -0.0 0.0" xyz="0.0 0.0 0.0"/>
            <geometry>
                <mesh filename="package://my_robot_arm/meshes/Test2_Base_Link_Body.dae"/>
            </geometry>
        </visual>
        <collision>
            <!--Base_Link_Part/Body.-->
            <origin rpy="0.0 -0.0 0.0" xyz="0.0 0.0 0.0"/>
            <geometry>
                <mesh filename="package://my_robot_arm/meshes/Test2_Base_Link_Body.dae"/>
            </geometry>
        </collision>
        <xacro:box_inertial_with_origin x="0.398" y="0.221" z="0.150" mass="1.0">
            <origin xyz="-0.15 0 0.03" rpy="0 0 0" />
        </xacro:box_inertial_with_origin>
    </link>
    <link name="finger_link">
        <visual>
            <!--finger_part/Body001.-->
            <origin rpy="0.0 -0.0 0.0" xyz="0.0 0.0 0.0"/>
            <geometry>
                <mesh filename="package://my_robot_arm/meshes/Test2_finger_body.dae"/>
            </geometry>
        </visual>
        <collision>
            <!--finger_part/Body001.-->
            <origin rpy="0.0 -0.0 0.0" xyz="0.0 0.0 0.0"/>
            <geometry>
                <mesh filename="package://my_robot_arm/meshes/Test2_finger_body.dae"/>
            </geometry>
        </collision>
        <xacro:box_inertial_with_origin x="0.398" y="0.221" z="0.150" mass="100.0">
            <origin xyz="-0.15 0 0.03" rpy="0 0 0" />
        </xacro:box_inertial_with_origin>
    </link>
    <joint name="finger_joint" type="prismatic">
        <parent link="base_link"/>
        <child link="finger_link"/>
        <origin rpy="0.0 -0.0 0.0" xyz="-0.06 -0.001 0.06"/>
        <axis xyz="1 0 0"/>
        <limit effort="2000.0" lower="-10.09" upper="10.078" velocity="10"/>
    </joint>

    <joint name="gripper_joint" type="fixed">
        <parent link="ur5_tool0"/>
        <child link="base_link"/>
    </joint>

    <ros2_control name="ur5_" type="system">
        <hardware>
            <plugin>gz_ros2_control/GazeboSimSystem</plugin>
        </hardware>

        <!-- <joint name="finger_joint">
            <command_interface name="position"/>
            <state_interface name="position">
                <param name="initial_value">0.0</param>
            </state_interface>
            <state_interface name="velocity"/>
        </joint> -->

        <joint name="ur5_shoulder_pan_joint">
            <command_interface name="position"/>
            <state_interface name="position">
                <param name="initial_value">0.0</param>
            </state_interface>
            <state_interface name="velocity"/>
        </joint>
        <joint name="ur5_shoulder_lift_joint">
            <command_interface name="position"/>
            <state_interface name="position">
                <param name="initial_value">0.0</param>
            </state_interface>
            <state_interface name="velocity"/>
        </joint>
        <joint name="ur5_elbow_joint">
            <command_interface name="position"/>
            <state_interface name="position">
                <param name="initial_value">0.0</param>
            </state_interface>
            <state_interface name="velocity"/>
        </joint>
        <joint name="ur5_wrist_1_joint">
            <command_interface name="position"/>
            <state_interface name="position">
                <param name="initial_value">0.0</param>
            </state_interface>
            <state_interface name="velocity"/>
        </joint>
        <joint name="ur5_wrist_2_joint">
            <command_interface name="position"/>
            <state_interface name="position">
                <param name="initial_value">0.0</param>
            </state_interface>
            <state_interface name="velocity"/>
        </joint>
        <joint name="ur5_wrist_3_joint">
            <command_interface name="position"/>
            <state_interface name="position">
                <param name="initial_value">0.0</param>
            </state_interface>
            <state_interface name="velocity"/>
        </joint>
        <joint name="finger_joint">
            <command_interface name="position"/>
            <state_interface name="position">
                <param name="initial_value">0.0</param>
            </state_interface>
            <state_interface name="velocity"/>
        </joint>


    </ros2_control>

    <gazebo>
        <plugin filename="gz_ros2_control-system" name="gz_ros2_control::GazeboSimROS2ControlPlugin">
            <parameters>$(find my_robot_arm)/config/ros2_control.yaml</parameters>

        </plugin>

        <plugin filename="gz-sim-sensors-system" name="gz::sim::systems::Sensors">
            <render_engine>ogre2</render_engine>
        </plugin>
    </gazebo>
</robot>